homeassistant/components/hdmi_cec.py

```diff
@@ -4,118 +4,535 @@
+                                                   range(15))))
+                if new_devices:
+                    discovery.load_platform(hass, dev_type, DOMAIN, discovered={ATTR_NEW: new_devices},
+                                            hass_config=base_config)
+                seconds_since_scan = 0
+                while seconds_since_scan < SCAN_INTERVAL:
+                    if stop.is_set():
+                        return
 
-        if _CEC.Open(adapters[0].strComName):
-            hass.services.register(DOMAIN, SERVICE_POWER_ON, _power_on)
-            hass.services.register(DOMAIN, SERVICE_STANDBY, _standby)
-            hass.services.register(DOMAIN, SERVICE_SELECT_DEVICE,
-                                   _select_device)
+                    time.sleep(1)
+                    seconds_since_scan += 1
         else:
-            _LOGGER.error("Failed to open adapter")
+            return False
+
+    def _stop_cec(event):
+        stop.set()
 
     hass.bus.listen_once(EVENT_HOMEASSISTANT_START, _start_cec)
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, _stop_cec)
     return True
+
+
+def icon_by_type(cec_type):
+    return ICON_TV if cec_type == 0 \
+        else ICON_RECORDER if cec_type == 1 \
+        else ICON_TUNER if cec_type == 3 \
+        else ICON_PLAYER if cec_type == 4 \
+        else ICON_AUDIO if cec_type == 5 \
+        else ICON_UNKNOWN
+
+
+class CecDevice(Entity):
+    """Representation of a HDMI CEC device entity."""
+
+    def __init__(self, hass, cec_client, logical):
+        """Initialize the device."""
+        self.hass = hass
+        self._icon = None
+        self._state = STATE_UNKNOWN
+        self._physical_address = None
+        self._logical_address = logical
+        self._vendor_id = None
+        self._cec_type_id = None
+        self._cec_type_id = CEC_LOGICAL_TO_TYPE[logical]
+        self._available = False
+        self._hidden = False
+        self._name = None
+        DEVICE_PRESENCE[logical] = True
+        _LOGGER.info("Initializing CEC device %s", self.name)
+        self.cec_client = cec_client
+        hass.bus.listen(EVENT_CEC_COMMAND_RECEIVED, self._update_callback)
+        self.entity_id = "%s.%d" % (DOMAIN, self._logical_address)
+
+    @asyncio.coroutine
+    def async_update(self):
+        yield from self.async_update_availability()
+        if self.available:
+            _LOGGER.info("Updating status for device %s", hex(self._logical_address)[2:])
+            yield from self.async_request_cec_power_status()
+            yield from self.async_request_cec_osd_name()
+            yield from self.async_request_cec_vendor()
+            yield from self.async_request_physical_address()
+        else:
+            _LOGGER.info("device not available. Not updating")
+
+    @asyncio.coroutine
+    def async_update_availability(self):
+        self._available = self.cec_client.lib_cec.PollDevice(self._logical_address)
```

Comment:
If you have a coroutine, all IO should also be handled via a coroutine, so if `PollDevice` was a coroutine, you could have done:

```
_available = yield PollDevice()
```

There are ways to schedule non coroutine IO functions, if you keep reading the async docs, but if none of your IO functions are coroutines, you should consider not using async for your component and hass will do all the heavy lifting.